### DISCUSSION TITLE
package-lock.json not updated due to 404-errors

### DISCUSSION THREAD

**Discussion by nlmarco** [renovatebot/renovate#18762] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.17.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

For security reasons, we deleted the `.npmrc`-files, which were previously checked-in into git, from all repos and replaced them by CI-vars. With the `.npmrc`-files inside the repos, it worked correctly. Now, we use a configuration like this:

````
module.exports = {
    ...
    hostRules: [
        ...,
        {
            matchHost: process.env.NPM_REGISTRY,
            token: process.env.NPM_AUTH,
            authType: 'Basic'
        }
    ],
    npmrc: 'registry=' + process.env.NPM_REGISTRY + '\nemail=' + process.env.NPM_EMAIL + '\nalways-auth=true\nloglevel=info'
};
````

We tried it with and without `authType` in the `hostRules`-entry for our Artifactory. We also tried it with and without `hostType: 'npm'`.

RenovateBot succeeds to update the `package.json`, but it fails to update the `package-lock.json`. There are tons of 404-errors which mean that authentication at the artifactory doesn't happen at all or fails:

```
DEBUG: rawExec err (repository=GAL/galileo.pos, branch=renovate/patch-nts-internal-automerge)
       "err": {
         "name": "ExecError",
         "cmd": "/bin/sh -c npm install --package-lock-only --no-audit --ignore-scripts",
         "stderr": "npm info using npm@8.19.3\nnpm info using node@v18.12.0\nnpm timing npm:load:whichnode Completed in 0ms\n ... ... ...\nnpm http fetch GET 404 https://**redacted**@adobe%2fcss-tools 25ms (cache skip)\nnpm http fetch GET 404 https://**redacted**@adobe%2fcss-tools 7ms (cache skip)\nnpm ... ... ...
```

Before we added the `hostRules`-entry for our artifactory and the `npmrc`-property, we copied the npmrc-contents from the temporary file whose path is made available by gitlab in the env-var `NPM_CONFIG_USERCONFIG` into `~/.npmrc`, but RenovateBot seems to completely ignore both. It does not care about the env-var (even though this is npm's standard-behaviour) and it does not care about the user-home's `.npmrc` either.

Without adding the `hostRules`-entry and the `npmrc`-property to RenovateBot's `config.js`, authentication at our artifactory fails completely. Even though the env-var `NPM_CONFIG_USERCONFIG` exists and an additional `~/.npmrc` exists, too. Why doesn't RenovateBot simply use the user's `.npmrc`??? Would be sooo easy.

After adding our Artifactory to the `hostRules` and adding the `npmrc`-property, RenovateBot got further and created merge-requests again -- but only for the `package.json`. Updating the `package-lock.json` isn't working no matter what variant of configuration we try.

This clearly is a bug of RenovateBot. If it succeeds updating the `package.json` correctly, it means that our configuration is correct. If it fails further down the route when running an `npm install` for the `package-lock.json`, it means that RenovateBot fails to properly apply its configuration to the child-process.

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
I don't think that this is a bug.  I know it's working with artifactory. Search discussion for samples.

Did you follow this[^1] guide?

[^1]: https://docs.renovatebot.com/getting-started/private-packages/#add-hostrule-to-bots-config

**Reply by nlmarco** [reply]
Yes, I followed this guide. That's why I wrote that I tried it with and without the "authType". It works better without it, i.e. it creates a merge-request for the `package.json` and only fails later at the `package-lock.json`.

Also I tried both types of token -- the one used with `_auth` and the one used with `_authToken`. In contrast to the guide you mentioned, it does not work with the one used with `_auth`. The documentation is contradictory, anyway:

The comment in the example code linked by you states:
```
// Will be passed as `//artifactory.my-company.com/artifactory/api/npm/npm:_auth=<TOKEN>` to `.npmrc`
```
But on the same page all other examples talk about `_authToken`.

And as said: It does not work with the short token used by `_auth`. It only works _partially_ with the long token used by `_authToken`. Hence, I think that the quoted comment is wrong.

**Reply by nlmarco** [reply]
And one more detail: I also tried adding the property `npmrcMerge` with value `true` to my `config.js`:

```
module.exports = {
    ...
    hostRules: [
        ...
    ],
    npmrcMerge: true,
```

...with all my settings (`registry`, `authToken` etc.) in the `$HOME/.npmrc` (and also in the temporary file generated by gitlab and referenced in the env-var `NPM_CONFIG_USERCONFIG`), but RenovateBot gives a shit about these `.npmrc`-files -- which btw. works fine (the env-var with the temporary file alone) with all our normal build-jobs, i.e. with `npm ci` fetching the libs from our Artifactory.

**Reply by nlmarco** [reply]
We found the bug:
```
module.exports = {
    ...
    npmrc: 'registry=' + process.env.NPM_REGISTRY + '\nemail=' + process.env.NPM_EMAIL + '\nalways-auth=true\nloglevel=info'
```
RenovateBot seems to incorrectly parse or process this `npmrc`-content. It works, if we pass nothing but the registry, i.e. no other options, like this:
```
module.exports = {
    ...
    hostRules: [
        {
            hostType: 'npm',
            matchHost: 'https://artifactory.company.com/npm/',
            token: '121212121212...long...token...for...npm..._authToken...183198723177123cnsdsdjfnjk'
        }
    ],
    npmrc: 'registry=https://artifactory.company.com/npm/'
```
The above, initial value was our real `.npmrc`-file's content with all line-breaks replaced by `\n`.

This is a bug. At the very least, it should be documented [here](https://docs.renovatebot.com/getting-started/private-packages/#add-npmrc-string-to-renovate-config) in red and bold that the `npmrc`-property must not include anything except for the registry. Better still would be to fix RenovateBot so that it correctly handles the contents of this property.

**Reply by rarkins** [reply]
No such limitation exists, to our knowledge. Maybe instead your env variable was missing the trailing forward slash which is present in your string literal. Try using the env again only for the registry part.

**Reply by nlmarco** [reply]
No. Our env-variable NPM_REGISTRY ends with a "/" for sure! I know this 100%, because I changed it this morning.

After my colleague tried it with hard-coded values, I replaced them by exactly the same env-vars as before -- and it works fine, now. The only difference between the version failing to update the `package-lock.json` due to authentication-errors and the current version working fine is that we removed all the stuff in the `npmrc`-value, except for the `registry` -- it now looks like this:
```
npmrc: 'registry=' + process.env.NPM_REGISTRY
```
Maybe it isn't a general problem relating to further entries in the `npmrc`, but one of our previously present additional options is causing the problem:
* `email=...`
* `always-auth=true`
* `loglevel=info`

Finally, a hint for those currently struggling: You must use the long token in your `hostRules`-entry, i.e. the one that you'd pass as `_authToken` in your `.npmrc`-file. It does *not* work with the short token (that would be passed as `_auth`-value).
